### PR TITLE
[Qol][Refactor] i18n lazy-loading

### PR DESCRIPTION
**Pull request by flx-sta**
> [!CAUTION]
> ### Make sure to consult the *@pagefaultgames/translation-team* & *@pagefaultgames/translation-leads*  before merging as all locale files moved from `/src/locales` into `/public/locales` !!!

> [!TIP]
> there's only `~7` relevant files, the rest are all file moves/etc
> - [@types/i18next.d.ts](https://github.com/pagefaultgames/pokerogue/pull/4327/files?file-filters%5B%5D=.ts&show-deleted-files=true&show-viewed-files=true#diff-d78a60876ad1873a6f1f26a69962182be27b51f07bb9606bb6209e6912ebf319)
> - [loading-scene.ts](https://github.com/pagefaultgames/pokerogue/pull/4327/files?file-filters%5B%5D=.ts&show-deleted-files=true&show-viewed-files=true#diff-99d6ceeb58f5b1d3a15a324e96b2a7311ef041cab023c56107c3e61badc4f8ea)
> - [main.ts](https://github.com/pagefaultgames/pokerogue/pull/4327/files?file-filters%5B%5D=.ts&show-deleted-files=true&show-viewed-files=true#diff-4fab5baaca5c14d2de62d8d2fceef376ddddcc8e9509d86cfa5643f51b89ce3d)
> - [plugins/i18n.ts](https://github.com/pagefaultgames/pokerogue/pull/4327/files?file-filters%5B%5D=.ts&show-deleted-files=true&show-viewed-files=true#diff-60258ac1f55717ab75876c0d8a946e2880a81c964b78b44ba15aa269861de544)
> - [test/utils/mocks/mocksContainer/mockText.ts](https://github.com/pagefaultgames/pokerogue/pull/4327/files?file-filters%5B%5D=.ts&show-deleted-files=true&show-viewed-files=true#diff-245f89be077f680057ab52be026cbcd710fc2f2b5f1eb099a2d8bbfba31337bd)
> - [test/vitest.setup.ts](https://github.com/pagefaultgames/pokerogue/pull/4327/files?file-filters%5B%5D=.ts&show-deleted-files=true&show-viewed-files=true#diff-7ccd55f6c0451a683566386d23e2ea3afbbd3fc41edb82816520318dc4d16740)
> - [utils.ts](https://github.com/pagefaultgames/pokerogue/pull/4327/files?file-filters%5B%5D=.ts&show-deleted-files=true&show-viewed-files=true#diff-39b2554fd18da165b59a6351b1aafff3714e2a80c1435f2de9706355b4d32351)
>
>  \- *NightKev*

---

## What are the changes the user will see?

Only the languages being used will be loaded thus less memory usage from the i18n side 

## Why am I making these changes?

The memory shouldn't be bloated with all languages at all time. IT should only hold the languages currently being used which at max is 2 (en as a fallback for missing keys).

## What are the changes from a developer perspective?

1. Moved all `/src/locales` into `/public/locales` as they will be loaded async (via network request).
2. Install the [i18next-http-backend](https://github.com/i18next/i18next-http-backend) plugin which is necessary for lazy-loading
3. Add a `camelCaseToKebabCase` function to utils
4. Clean up `/src/plugin/i18n.ts` file structure
5. move phaser config into `startGame()` and lazy-load the scenes to prevent any `i18n.t()` usages before it is initialized
6. Remove `initI18n()` from `LoadingScene`
7. Remove all `config.ts` files inside the `locales/` 
8. **Remove all `locales` tests**. _Reason: We don't need to test if i18next is working properly. They have their own test-coverage to ensure that_
9. **`i18next` is now mocked for tests**. <sub>This was overdue and now necessary due to the backend module</sub>

## How to test the changes?

1. Play the game
2. Switch languages and keep on playing.
3. Check the network tab for any failed translation requests (`locales/**`)

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- ~~[ ] Have I considered writing automated tests for the issue?~~
- ~~[ ] If I have text, did I make it translatable and add a key in the English locale file(s)?~~
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- ~~[ ] Are the changes visual?~~
  - ~~[ ] Have I provided screenshots/videos of the changes?~~

### CODE REVIEW

**Comment by flx-sta** [comment]
# Review Helper

to mark all `**/*.json` files as `Viewed` in the `Files` tab simply run this script in your browser-console:
```js
 $$("#files .js-reviewed-toggle > input[id*='.json']").forEach((el) => el.click()); 
 ```

File: src/test/vitest.setup.ts [hunk 1]
```diff
@@ -20,10 +20,31 @@ vi.mock("#app/overrides", async (importOriginal) => {
...
 });
 
+vi.mock("i18next", () => ({
+  default: {
+    use: () => {},
+    t: (key: string) => key,
+    changeLanguage: () => Promise.resolve(),
+    init: () => Promise.resolve(),
+    resolvedLanguage: "en",
+    exists: vi.fn(() => true),
```
Comment (DayKev) [inline src/test/vitest.setup.ts:34]: Is the `vi.fn()` usage necessary? As discovered before, every new `vi.fn` adds a memory leak (can't remember if it also increased test time).

File: src/test/vitest.setup.ts [hunk 2]
```diff
@@ -20,10 +20,31 @@ vi.mock("#app/overrides", async (importOriginal) => {
...
+    getDataByLanguage:() => ({
+      en: {
+        keys: ["foo"]
+      },
+    }),
+    services: {
+      formatter: {
+        add: vi.fn(),
```
Comment (DayKev) [inline src/test/vitest.setup.ts:42]: (and here)

Comment (flx-sta) [reply src/test/vitest.setup.ts:42]: nope

File: src/test/vitest.setup.ts (hunk 1 above)
Comment (flx-sta) [reply src/test/vitest.setup.ts:34]: nope

**Review by Chapybara-jp — APPROVED** [review]

**Review by Chapybara-jp — DISMISSED** [review]

File: src/@types/i18next.d.ts [hunk 3]
```diff
@@ -1,12 +1,11 @@
-import { type enConfig } from "#app/locales/en/config";
 import { TOptions } from "i18next";
 
 //TODO: this needs to be type properly in the future
 // Module declared to make referencing keys in the localization files type-safe.
 declare module "i18next" {
     interface CustomTypeOptions {
       defaultNS: "menu", // needed here as well for typedoc
-      resources: typeof enConfig
+      // resources: typeof enConfig
```
Comment (ben-lear) [inline src/@types/i18next.d.ts:8]: is this meant to stay commented?

**Comment by DayKev** [comment]
> # Review Helper
> to mark all `**/*.json` files as `Viewed` in the `Files` tab simply run this script in your browser-console:
> 
> ```js
>  $$("#files .js-reviewed-toggle > input[id*='.json']").forEach((el) => el.click()); 
> ```
> 
> Your browser will probably lagg for 1-3 minutes as it sends quite a lot of server-requests at the same time then

An alternative to this is using the file filter and disabling `.json` files (ie, click this: https://github.com/pagefaultgames/pokerogue/pull/4327/files?file-filters%5B%5D=.ts&show-deleted-files=true&show-viewed-files=true)

File: src/@types/i18next.d.ts (hunk 3 above)
Comment (DayKev) [inline src/@types/i18next.d.ts:8]:
```suggestion
```

File: src/plugins/i18n.ts [hunk 4]
```diff
@@ -89,6 +100,12 @@ async function initFonts(language: string | undefined) {
   }
 }
 
+//#region Exports
+
+/**
+ * Initialize i18n with fonts
+ * @returns {Promise<void>}
```
Comment (DayKev) [inline src/plugins/i18n.ts:107]:
```suggestion
 * @returns {Promise<void>}
```

File: src/@types/i18next.d.ts (hunk 3 above)
Comment (flx-sta) [reply src/@types/i18next.d.ts:8]: no. whops

**Review by frutescens — DISMISSED** [review]

**Review by innerthunder — APPROVED** [review]

**Comment by MokaStitcher** [comment]
Found a missing change from `.` to`:` in one ME by pure chance when playtesting. 

https://github.com/pagefaultgames/pokerogue/blob/c287584bf1fb872bacc9c0ef4bb25712fe477544/src/data/mystery-encounters/encounters/the-expert-pokemon-breeder-encounter.ts#L543

Screenshot 2024-10-01 at 21 50 43

Will try and double check all ME files

**Review by CodeTappert — DISMISSED** [review]

**Review by MokaStitcher — DISMISSED** [review]
Found 2 keys not properly updated for MEs:

https://github.com/pagefaultgames/pokerogue/blob/2de15f9be29acd5f0fa678eb46a3d404faabbe8d/src/data/mystery-encounters/encounters/mysterious-chest-encounter.ts#L176
should be  `await showEncounterText(scene, `${namespace}:option.1.bad`); `


and

https://github.com/pagefaultgames/pokerogue/blob/c287584bf1fb872bacc9c0ef4bb25712fe477544/src/data/mystery-encounters/encounters/the-expert-pokemon-breeder-encounter.ts#L543

should be `text: `${namespace}:outro_failed`, `

**Review by frutescens — APPROVED** [review]

**Review by ben-lear — APPROVED** [review]

**Review by MokaStitcher — APPROVED** [review]

**Review by f-fsantos — APPROVED** [review]

**Pull request merged**